core/src/main/scala/com/github/johnynek/paiges/Doc.scala

```diff
@@ -110,6 +112,18 @@ sealed abstract class Doc extends Serializable {
   def writeTo(maxLine: Int, pw: PrintWriter): Unit =
     Doc.write(this, maxLine, pw)
 
+  override lazy val hashCode: Int = {
+    import Doc.Tok
+    @inline def hash(curr: Int, c: Char): Int =
+      curr * 1500450271 + c.toInt
+    @tailrec def shash(n: Int, s: String, i: Int): Int =
+      if (i < s.length) shash(hash(n, s.charAt(i)), s, i + 1) else n
+    Tok.fromDoc(this).foldLeft(0xdead60d5) {
+      case (n, Tok.Line(i)) => hash(n, '\n') * i
```

Comment:
might want `hash(n, '\n') * (i + 1)` or something since `i == 0` is the inner most union, maybe most common?

Actually, why not: `1500450271  * i + hash(n, '\n')`